storage/src/main/java/org/apache/uniffle/storage/handler/impl/PrefetchableClientReadHandler.java

```diff
@@ -0,0 +1,136 @@
+  private ExecutorService prefetchExecutors;
+  private AtomicBoolean abnormalFetchTag;
+  private AtomicBoolean finishedTag;
+
+  public PrefetchableClientReadHandler(Optional<PrefetchOption> prefetchOptional) {
+    if (prefetchOptional.isPresent()) {
+      PrefetchOption option = prefetchOptional.get();
+      if (option.capacity <= 0) {
+        throw new RssException("Illegal prefetch capacity: " + option.capacity);
+      }
+      LOG.info("Prefetch is enabled, capacity: {}", option.capacity);
+      this.prefetchEnabled = true;
+      this.prefetchQueueCapacity = option.capacity;
+      this.prefetchTimeoutSec = option.timeoutSec;
+      this.prefetchResultQueue = new LinkedBlockingQueue<>(option.capacity);
+      // todo: support multi threads to prefetch
+      this.prefetchExecutors = Executors.newFixedThreadPool(1);
+      this.abnormalFetchTag = new AtomicBoolean(false);
+      this.finishedTag = new AtomicBoolean(false);
+    } else {
+      this.prefetchEnabled = false;
+    }
+  }
+
+  public static class PrefetchOption {
+    private int capacity;
+    private int timeoutSec;
+
+    public PrefetchOption(int capacity, int timeoutSec) {
+      this.capacity = capacity;
+      this.timeoutSec = timeoutSec;
+    }
+  }
+
+  protected abstract ShuffleDataResult doReadShuffleData();
+
+  @Override
+  public ShuffleDataResult readShuffleData() {
+    if (!prefetchEnabled) {
+      return doReadShuffleData();
+    }
+
+    int free = prefetchQueueCapacity - prefetchResultQueue.size();
+    for (int i = 0; i < free; i++) {
+      prefetchExecutors.submit(
+          () -> {
+            // if it has been marked as abnormal/finished state, skip the following fetching.
+            if (abnormalFetchTag.get() || finishedTag.get()) {
+              return;
+            }
+            try {
+              ShuffleDataResult result = doReadShuffleData();
+              if (result == null) {
+                this.finishedTag.set(true);
+              }
+              prefetchResultQueue.offer(Optional.ofNullable(result));
+            } catch (Exception e) {
+              abnormalFetchTag.set(true);
+              LOG.error("Errors on doing readShuffleData", e);
+            }
+          });
+    }
+
+    long start = System.currentTimeMillis();
+    while (true) {
+      if (abnormalFetchTag.get()) {
+        throw new RssException("Fast fail due to the fetch failure");
+      }
+
+      Optional<ShuffleDataResult> optionalShuffleDataResult = prefetchResultQueue.poll();
+      if (optionalShuffleDataResult != null) {
+        if (optionalShuffleDataResult.isPresent()) {
+          return optionalShuffleDataResult.get();
+        } else {
+          return null;
+        }
+      }
+
+      if (System.currentTimeMillis() - start > prefetchTimeoutSec * 1000) {
+        throw new RssException("Unexpected duration of reading shuffle data. Fast fail!");
+      }
+
+      try {
+        Thread.sleep(10);
```

Comment:
Method `wait` may be better than sleep.

Reply:
Wait is OK but introduce extra sync code. Let me take a deep think

Reply:
Sleep will occupy the cpu and wait won't occupy the cpu.

Reply:
Removed